Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,354 @@
+            return UniqueEvents.Count + SessionEvents.Count;
+        }
+
+        /// <summary>
+        /// Add session event to next Backtrace session request
+        /// </summary>
+        /// <param name="eventName">session event name</param>
+        public bool AddSessionEvent(string eventName, IDictionary<string, string> attributes = null)
+        {
+            if (!ShouldProcessEvent(eventName))
+            {
+                Debug.LogWarning("Skipping report: Reached store limit or event has empty name.");
+                return false;
+            }
+
+            var @event = new SessionEvent(eventName, DateTimeHelper.Timestamp(), attributes);
+            SessionEvents.AddLast(@event);
+            return true;
+        }
+
+        /// <summary>
+        /// Clean unique and session events stored in Backtrace Session
+        /// </summary>
+        public void ClearEvents()
+        {
+            UniqueEvents.Clear();
+            SessionEvents.Clear();
+        }
+
+        private void OnRequestCompleted()
+        {
+            _numberOfDroppedRequests = 0;
+            return;
+        }
+
+        private BacktraceJObject CreateJsonPayload(ICollection<UniqueEvent> uniqueEvents, ICollection<SessionEvent> sessionEvents)
+        {
+            var jsonData = new BacktraceJObject();
+            jsonData.Add("application", Application.productName);
+            jsonData.Add("appversion", Application.version);
+            jsonData.Add("metadata", CreatePayloadMetadata());
+
+            // add unique events
+            var uniqueEventsJson = new List<BacktraceJObject>();
+            foreach (var uniqueEvent in uniqueEvents)
+            {
+                uniqueEventsJson.Add(uniqueEvent.ToJson());
+            }
+
+            jsonData.Add("unique_events", uniqueEventsJson);
+
+            // add session events
+            var sessionEventsJson = new List<BacktraceJObject>();
+            var attributes = _attributeProvider.Get();
+            foreach (var sessionEvent in sessionEvents)
+            {
+                sessionEventsJson.Add(sessionEvent.ToJson(attributes));
+            }
+
+            jsonData.Add("session_events", sessionEventsJson);
+            return jsonData;
+        }
+
+        private BacktraceJObject CreatePayloadMetadata()
+        {
+            var payload = new BacktraceJObject();
+            payload.Add("dropped_events", _numberOfDroppedRequests);
+            return payload;
+        }
+
+        private double CalculateNextRetryTime(uint numberOfRetries)
+        {
+            const int jitterFraction = 1;
+            const int backoffBase = 10;
+            var value = DefaultTimeInSecBetweenRequests * Math.Pow(backoffBase, numberOfRetries);
+            var retryLower = Math.Max(0, Math.Min(RetryTimeMax, value));
+            var retryUpper = retryLower + retryLower * jitterFraction;
+            return new System.Random().NextDouble() * (retryUpper - retryLower) + retryLower;

```

Comment:
Fine if we don't have one, but this would be clearer if there were a version of random that took the endpoints of the range, rather than requiring us to scale.

Reply:
Unity uniform function require to use floats. Here we have doubles. So I prefer to stay with this implementation instead of just casting it to floats.

Reply:
Moved to Internal helpers (like Clamp)